File: elixir/resistor-color-duo/lib/resistor_color_duo.ex
defmodule ResistorColorDuo do
  @moduledoc """
  Helping with color-coded resistors.
  """

  @colors ~w(black brown red orange yellow green blue violet grey white)a

  @doc """
  Calculate a resistance value from two colors
  """
  @spec value(colors :: [atom]) :: integer
  def value(colors) do
    String.to_integer("#{get_color(Enum.at(colors, 0))}#{get_color(Enum.at(colors, 1))}")
  end

  @spec get_color(color :: atom) :: integer
  defp get_color(color) do
    Enum.find_index(@colors, fn c -> c == color end)
  end
end

File: typescript/resistor-color-duo/resistor-color-duo.ts
const COLORS : { [color: string]: string } = {
  'black': '0',
  'brown': '1',
  'red': '2',
  'orange': '3',
  'yellow': '4',
  'green': '5',
  'blue': '6',
  'violet': '7',
  'grey': '8',
  'white': '9'
};

export function decodedValue(colorCoding : string[]) : number {
  const stringifiedValue : string =
    colorCoding
      .slice(0, 2)
      .map((color : string) => COLORS[color.toLocaleLowerCase()])
      .reduce((value, sum) => value + sum, '');

  return parseInt(stringifiedValue, 10);
}
